File: implementations/java/osf-java/src/test/java/com/optimeas/osf/ConformanceManifestTest.java
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Optimeas GmbH
package com.optimeas.osf;

import com.optimeas.osf.testutil.ExamplesDir;
import com.optimeas.osf.testutil.ReferenceManifest;
import com.optimeas.osf.testutil.ReferenceManifest.ChannelEntry;
import com.optimeas.osf.testutil.ReferenceManifest.FileEntry;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import java.io.IOException;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Cross-implementation conformance test: asserts that {@link DataManager}
 * decodes every generated reference file to match the expected contents
 * declared in {@code examples/reference_manifest.json}.
 *
 * <p>Each parameterized invocation loads one file from {@code examples/generated/}
 * and verifies:
 * <ul>
 *   <li>The declared manifest version matches the OSF magic header (inferred
 *       from the canonical filename prefix {@code osf4_} / {@code osf5_}).</li>
 *   <li>Channel count matches the manifest.</li>
 *   <li>Per channel (matched by index): name, data type (wire name), sample
 *       count, and storage mode ({@code equidistant} / {@code timestamped} /
 *       {@code variable}).</li>
 * </ul>
 *
 * <p>The test skips cleanly when the corpus or manifest is absent (via
 * {@link ExamplesDir}).
 */
class ConformanceManifestTest {

    /**
     * Provide one argument per manifest entry: the resolved path to
     * {@code examples/generated/<file>}.
     *
     * <p>If the examples directory is absent {@code ExamplesDir.resolve()}
     * causes the assumption to fail and the whole test is skipped. If the
     * directory exists but the generated sub-directory is empty the stream
     * is empty and the parameterized test reports zero invocations.
     */
    static Stream<String> manifestKeys() throws IOException {
        // ExamplesDir.resolve() skips if corpus absent.
        ExamplesDir.resolve();
        return ReferenceManifest.load().keySet().stream().sorted();
    }

    @ParameterizedTest(name = "{0}")
    @MethodSource("manifestKeys")
    void conformsToManifest(String key) throws IOException {
        // The manifest key may be a sub-path (e.g.
        // "integrity/osf5_crc_equidistant.osf") — those files live in their own
        // directory but are still driven by the shared manifest. Resolve the
        // file from the key, and look the entry up by the full key (not the bare
        // filename, which would not match a sub-path key).
        Path file = ExamplesDir.resolve().resolve("generated").resolve(key);
        String fileName = file.getFileName().toString();

        Map<String, FileEntry> manifest = ReferenceManifest.load();
        FileEntry expected = manifest.get(key);
        assertThat(expected)
                .as("manifest entry for %s", key)
                .isNotNull();

        // ── Version check ────────────────────────────────────────────────────
        // DataManager does not expose OsfVersion publicly; we infer the
        // expected format version from the canonical filename prefix and assert
        // it matches the manifest declaration.  Correct decoding (channels /
        // types / counts below) would fail independently if the wrong parser
        // were chosen.
        int versionFromName = fileName.startsWith("osf4_") ? 4 : 5;
        assertThat(expected.version())
                .as("manifest version for %s matches filename prefix", key)
                .isEqualTo(versionFromName);

        // ── Load ────────────────────────────────────────────────────────────
        DataManager mgr = DataManager.loadFromFile(file);

        // ── Integrity profile (optional) ─────────────────────────────────────
        // Entries that declare an integrity profile must decode with that
        // profile detected and every block's frame CRC verified.
        if (expected.integrity() != null) {
            IntegrityProfile expectedProfile = switch (expected.integrity()) {
                case "crc32c" -> IntegrityProfile.CRC32C;
                case "ed25519" -> IntegrityProfile.ED25519;
                default -> IntegrityProfile.NONE;
            };
            assertThat(mgr.stats().integrity())
                    .as("integrity profile of %s", key)
                    .isEqualTo(expectedProfile);
            assertThat(mgr.stats().blocksCrcFailed())
                    .as("frame-CRC failures in %s", key)
                    .isZero();
        }

        // ── Anomalies (optional) ─────────────────────────────────────────────
        // Deliberate non-conformances (optional). A file that declares none
        // must report none — a well-formed file reporting a zero-length skip
        // is itself a finding. Asserted unconditionally, unlike the integrity
        // block above: a file with no integrity profile has no frame CRCs to
        // fail, but any file at all can carry a zero-length block.
        long wantZeroLength = expected.anomalies().getOrDefault("zeroLengthBlocks", 0L);
        assertThat(mgr.stats().blocksSkippedZeroLength())
                .as("%s: anomalies.zeroLengthBlocks", key)
                .isEqualTo(wantZeroLength);

        List<DataChannel> channels = mgr.channels();

        // ── Channel count ────────────────────────────────────────────────────
        assertThat(channels)
                .as("channel count in %s", fileName)
                .hasSize(expected.channels().size());

        // ── Per-channel assertions ───────────────────────────────────────────
        for (ChannelEntry ce : expected.channels()) {
            DataChannel ch = channels.stream()
                    .filter(c -> c.index() == ce.index())
                    .findFirst()
                    .orElse(null);

            assertThat(ch)
                    .as("channel at index %d in %s", ce.index(), fileName)
                    .isNotNull();

            assertThat(ch.name())
                    .as("name of channel %d in %s", ce.index(), fileName)
                    .isEqualTo(ce.name());

            assertThat(ch.dataType().wireName())
                    .as("dataType wireName of channel %d (%s) in %s",
                            ce.index(), ce.name(), fileName)
                    .isEqualTo(ce.dataType());

            assertThat(ch.sampleCount())
                    .as("sampleCount of channel %d (%s) in %s",
                            ce.index(), ce.name(), fileName)
                    .isEqualTo(ce.sampleCount());

            String expectedMode = modeFromKind(ch.kind());
            assertThat(expectedMode)
                    .as("mode of channel %d (%s) in %s",
                            ce.index(), ce.name(), fileName)
                    .isEqualTo(ce.mode());
        }
    }

    /**
     * Map {@link DataChannel.Kind} to the manifest's {@code mode} string.
     *
     * @param kind the storage kind
     * @return {@code "equidistant"}, {@code "timestamped"}, or {@code "variable"}
     * @throws IllegalArgumentException for any unrecognised future kind
     */
    private static String modeFromKind(DataChannel.Kind kind) {
        return switch (kind) {
            case EQUIDISTANT -> "equidistant";
            case TIMESTAMPED -> "timestamped";
            case VARIABLE    -> "variable";
        };
    }
}

File: implementations/rust/osf-core/tests/conformance_manifest_test.rs
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Optimeas GmbH

//! Manifest-driven cross-implementation conformance test.
//!
//! Reads the shared `examples/reference_manifest.json` — the single source of
//! truth for the expected decoded contents of every reference file — and asserts
//! that `DataManager` decodes each listed file to match: version (via the
//! filename prefix), channel count, and per-channel index/name/datatype/
//! sample-count/mode, plus the integrity profile for entries that declare one,
//! plus the deliberate-anomaly counts (currently `zeroLengthBlocks`) declared
//! under the optional `anomalies` field. The anomaly assertion runs on every
//! entry, not only those with `anomalies` present: absent means the file is
//! well-formed and must report a zero count, present means the stated count is
//! exact — so a well-formed file that unexpectedly reports a skip is itself a
//! finding, not a silently-ignored corner case.
//!
//! Manifest keys may be sub-paths (e.g. `integrity/osf5_crc_equidistant.osf`);
//! they resolve under `examples/generated/`. Keeping this list in the manifest
//! (rather than hard-coded here) is what makes it a genuine cross-language
//! contract shared with the Java/C++/Delphi conformance tests.

use osf_core::{Channel, DataManager, DataType, IntegrityProfile, parse_data_type};
use serde::Deserialize;
use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

fn examples_root() -> PathBuf {
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .parent()
        .and_then(Path::parent)
        .and_then(Path::parent)
        .expect("crate is rooted at .../implementations/rust/osf-core")
        .join("examples")
}

#[derive(Deserialize)]
struct FileEntry {
    version: u32,
    #[serde(default)]
    integrity: Option<String>,
    #[serde(default)]
    anomalies: Option<Anomalies>,
    channels: Vec<ChannelEntry>,
}

/// Deliberate non-conformances a corpus file carries. Absent for well-formed
/// files. Each field is the exact count a conforming reader must report.
#[derive(Deserialize)]
struct Anomalies {
    // No `#[serde(default)]` here, deliberately: once `anomalies` is present
    // in an entry, `zeroLengthBlocks` is required. Defaulting would let a
    // typo'd key silently deserialise to 0 and pass, which matters once
    // other-language ports re-spell this key by hand without yet having a
    // reader counter behind it.
    #[serde(rename = "zeroLengthBlocks")]
    zero_length_blocks: u64,
}

#[derive(Deserialize)]
struct ChannelEntry {
    index: u16,
    name: String,
    #[serde(rename = "dataType")]
    data_type: String,
    #[serde(rename = "sampleCount")]
    sample_count: usize,
    mode: String,
}

fn mode_of(ch: &Channel) -> &'static str {
    match ch {
        Channel::Equidistant(_) => "equidistant",
        Channel::Timestamped(_) => "timestamped",
        Channel::Variable(_) => "variable",
    }
}

fn integrity_of(token: &str) -> IntegrityProfile {
    match token {
        "crc32c" => IntegrityProfile::Crc32c,
        "ed25519" => IntegrityProfile::Ed25519,
        _ => IntegrityProfile::None,
    }
}

#[test]
fn conforms_to_reference_manifest() {
    let root = examples_root();
    let generated = root.join("generated");
    let manifest_path = root.join("reference_manifest.json");
    let text = std::fs::read_to_string(&manifest_path)
        .unwrap_or_else(|e| panic!("read {}: {e}", manifest_path.display()));
    let manifest: BTreeMap<String, FileEntry> =
        serde_json::from_str(&text).expect("reference_manifest.json is valid");
    assert!(!manifest.is_empty());

    for (key, entry) in &manifest {
        let path = generated.join(key);
        let mgr = DataManager::load_from_file(&path)
            .unwrap_or_else(|e| panic!("load {key}: {e}"));

        // Version inferred from the filename prefix.
        let base = Path::new(key).file_name().unwrap().to_str().unwrap();
        let want_version = if base.starts_with("osf4_") { 4 } else { 5 };
        assert_eq!(entry.version, want_version, "{key}: manifest version");

        // Integrity profile (optional).
        if let Some(token) = &entry.integrity {
            assert_eq!(mgr.stats.integrity, integrity_of(token), "{key}: integrity");
            assert_eq!(mgr.stats.blocks_crc_failed, 0, "{key}: crc failures");
        }

        // Deliberate non-conformances (optional). A file that declares none
        // must report none — that is what keeps a well-formed corpus honest.
        // Unlike the integrity check above (only asserted when `integrity` is
        // declared, since a file with no integrity profile has no frame CRCs
        // to fail), this runs unconditionally: any file at all, declared or
        // not, can carry a zero-length block, so "not declared" must mean
        // zero rather than "not checked".
        let want_zero_len = entry.anomalies.as_ref().map_or(0, |a| a.zero_length_blocks);
        assert_eq!(
            mgr.stats.blocks_skipped_zero_length, want_zero_len,
            "{key}: anomalies.zeroLengthBlocks (left = reader, right = manifest)"
        );

        let channels = mgr.channels();
        assert_eq!(channels.len(), entry.channels.len(), "{key}: channel count");

        for ce in &entry.channels {
            let ch = channels
                .iter()
                .find(|c| c.index() == ce.index)
                .unwrap_or_else(|| panic!("{key}: channel index {}", ce.index));
            assert_eq!(ch.name(), ce.name, "{key}: name of channel {}", ce.index);
            let want_dt: DataType = parse_data_type(&ce.data_type)
                .unwrap_or_else(|e| panic!("{key}: bad manifest dataType {:?}: {e}", ce.data_type));
            assert_eq!(ch.data_type(), want_dt, "{key}: dataType of channel {}", ce.index);
            assert_eq!(ch.sample_count(), ce.sample_count, "{key}: sampleCount of channel {}", ce.index);
            assert_eq!(mode_of(ch), ce.mode, "{key}: mode of channel {}", ce.index);
        }
    }
}
